pbft/pbft.cpp

```diff
@@ -0,0 +1,360 @@
+// Copyright (C) 2018 Bluzelle
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License, version 3,
+// as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+#include <pbft/pbft.hpp>
+#include <utils/make_endpoint.hpp>
+#include <google/protobuf/text_format.h>
+#include <cstdint>
+#include <memory>
+#include <boost/beast/core/detail/base64.hpp>
+#include <algorithm>
+
+using namespace bzn;
+
+pbft::pbft(
+        std::shared_ptr<bzn::node_base> node
+        , const bzn::peers_list_t& peers
+        , bzn::uuid_t uuid
+        , std::shared_ptr<pbft_service_base> service
+)
+        : node(std::move(node))
+        , peers(std::move(peers))
+        , uuid(std::move(uuid))
+        , service(service)
+{
+    if (this->peers.empty())
+    {
+        throw std::runtime_error("No peers found!");
+    }
+
+    std::transform(
+            this->peers.begin(), this->peers.end(), std::back_inserter(this->peer_index), [](auto& peer)
+            { return std::reference_wrapper<const peer_address_t>{peer}; }
+    );
+
+    std::copy(peers.begin(), peers.end(), std::back_inserter(peer_index));
+    std::sort(
+            peer_index.begin(), peer_index.end(), [](const auto& peer1, const auto& peer2)
+            { return peer1.get().uuid < peer2.get().uuid; }
+    );
+
+}
+
+void
+pbft::start()
```

Comment:
Isn't this only called from main?

Reply:
Then wouldn't it be better to throw an error, rather than silently ignoring subsequent calls?

Reply:
Yeah, that's my thinking. If we accidentally try to start one of these modules twice, then that's likely something very and obviously wrong - and it'd be easier to debug if it was an obvious crash rather than silently ignored at runtime.